FlowCal/plot.py

```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        if isinstance(y_bin_edges, collectionsAbc.Mapping):
+            try:
+                violin_y_bin_edges = y_bin_edges[violin_position]
+            except KeyError:
+                try:
+                    violin_y_bin_edges = y_bin_edges['data']
+                except KeyError:
+                    msg  = "unable to understand `y_bin_edges`"
+                    raise ValueError(msg)
+        else:
+            # check for sequence of sequences
+            try:
+                first_element = next(iter(y_bin_edges))
+                try:
+                    iter(first_element)   # success => sequence of sequences
+                    violin_y_bin_edges = y_bin_edges[idx]
+                except TypeError:
+                    violin_y_bin_edges = y_bin_edges
+            except TypeError:
+                msg  = "`y_bin_edges` should be iterable sequence or sequence"
+                msg += " of sequences"
+                raise TypeError(msg)
+
+        # understand upper and lower trim fractions
+        if isinstance(upper_trim_fraction, collectionsAbc.Mapping):
+            try:
+                v_upper_trim_fraction = upper_trim_fraction[violin_position]
+            except KeyError:
+                try:
+                    v_upper_trim_fraction = upper_trim_fraction['data']
+                except KeyError:
+                    msg  = "unable to understand `upper_trim_fraction`"
+                    raise ValueError(msg)
+        elif isinstance(upper_trim_fraction, collectionsAbc.Sequence):
+            v_upper_trim_fraction = upper_trim_fraction[idx]
+        else:
+            v_upper_trim_fraction = upper_trim_fraction
+
+        if isinstance(lower_trim_fraction, collectionsAbc.Mapping):
+            try:
+                v_lower_trim_fraction = lower_trim_fraction[violin_position]
+            except KeyError:
+                try:
+                    v_lower_trim_fraction = lower_trim_fraction['data']
+                except KeyError:
+                    msg  = "unable to understand `lower_trim_fraction`"
+                    raise ValueError(msg)
+        elif isinstance(lower_trim_fraction, collectionsAbc.Sequence):
+            v_lower_trim_fraction = lower_trim_fraction[idx]
+        else:
+            v_lower_trim_fraction = lower_trim_fraction
+
+        _plot_violin(violin_position=violin_position,
+                     violin_data=violin_data,
+                     violin_width=violin_width,
+                     violin_kwargs=v_kwargs,
+                     y_bin_edges=violin_y_bin_edges,
+                     xscale=xscale,
+                     upper_trim_fraction=v_upper_trim_fraction,
+                     lower_trim_fraction=v_lower_trim_fraction,
+                     draw_summary_stat=draw_summary_stat,
+                     draw_summary_stat_fxn=draw_summary_stat_fxn,
+                     draw_summary_stat_kwargs=v_draw_summary_stat_kwargs)
+
+        if draw_model:
```

Comment:
Why is this inside the `for idx in range(data_length):`? Does this mean that the model plot is made every time a new violin is plotted?

Reply:
Hmm, good catch. Yeah, that should probably be unindented once.